### ISSUE TITLE
unable to decrypt message on new but verified user session

### ISSUE THREAD

**Issue by CybotTM** [matrix-org/matrix-hookshot#974]
Messages sent by the Hookshot bot are consistently unreadable by some users due to decryption issues, displaying only "Unable to decrypt message." It can take up to two days before affected users can successfully decrypt and read the messages.

This issue occurs exclusively with messages from the Hookshot bot.

Is there a way to keep the bot online for an extended period to ensure successful key exchange?

Steps to reproduce:

- User: Having a active session.
- Inviting Hookshot Bot and create web hook.
- Web Hook: sends messages: All Fine
- User: (delete session or app and/or) creates a new session and verifies this session
- Web Hook: sends messages: User always get's "unable to decrypt message"

![image](https://github.com/user-attachments/assets/91bd7dbb-b3a3-4cff-9a43-88d434dcfffc)

Decrypted event:

```json
{
  "type": "m.room.message",
  "content": {
    "msgtype": "m.bad.encrypted",
    "body": "** Unable to decrypt: DecryptionError: The sender's device has not sent us the keys for this message. **"
  }
}
```

original event:
```json
{
  "type": "m.room.encrypted",
  "sender": "@_webhookhs_prometheus:netresearch.de",
  "content": {
    "algorithm": "m.megolm.v1.aes-sha2",
    "ciphertext": "Awg...OpAY",
    "device_id": "BOPKPMWUPZ",
    "sender_key": "pXN***N0Q",
    "session_id": "iFN***sOQ"
  },
  "origin_server_ts": 1731919570589,
  "unsigned": {
    "membership": "join",
    "age": 62
  },
  "event_id": "$Ht***-a4A",
  "room_id": "!***:netresearch.de"
}
```

Fix: drop web hook and recreate a new hook. User is able to read messages again ... until he is required to use/create a new session.

**Comment by CybotTM** [comment]
Problem seems solved with release 6.0.0

Thank You.

**Comment by PseudoResonance** [comment]
I'm fairly certain it is not solved.

**Comment by PseudoResonance** [comment]
> this happens to me whenever I connect a new verified device (e.g. a new browser). I can see messages in Rooms, but all my previous chats in People are just "unable to decrypt message". But I guess that's nothing to do with Hookshot so I should probably report this else where, I guess - it's just this issue came up in my searches.

From my very poor understanding, this is an issue with the "client" that is sending the messages. When a new device is added (or a new user joins), the sending client must listen for these events, and share their outbound Megolm keys with the new device. 

Because of the way Megolm keys work, once these initial keys are shared, the new device can decrypt all future messages, but not past messages. I guess that's why they call it the "[Megolm ratchet](https://matrix.org/docs/matrix-concepts/end-to-end-encryption/)," because it goes one way.

TLDR: Your issue by design and there is no fix. This issue is unrelated. (Again, this is just my very rough understanding though. I apologize if I terribly misunderstood...)

**Comment by CybotTM** [comment]
@PseudoResonance, could you please share your thoughts on why you believe this issue remains unresolved?

In our experience, it was resolved after we applied the update back then.

**Comment by PseudoResonance** [comment]
> [@PseudoResonance](https://github.com/PseudoResonance), could you please share your thoughts on why you believe this issue remains unresolved?

I mean, I'm glad to hear the issue seems to have gone away for you, but I'm not really sure why? Unless I'm missing something, nothing has been changed in the code that should resolve this, and unless I've done something wrong, I'm still getting the same issue.

Environment:
ghcr.io/matrix-org/matrix-hookshot:6.0.1
Redis with persistent storage
`passFile` in persistent storage

<details>
<summary>Hookshot Config</summary>

```yaml
bridge:
  bindAddress: 0.0.0.0
  domain: MATRIX-DOMAIN
  mediaUrl: https://MATRIX-DOMAIN
  port: 9002
  url: https://MATRIX-DOMAIN
listeners:
  - bindAddress: 0.0.0.0
	port: 9000
	resources:
	  - webhooks
	  - widgets
  - bindAddress: 127.0.0.1
	port: 9001
	resources:
	  - metrics
	  - provisioning
generic:
  enabled: true
  enableHttpGet: false
  urlPrefix: https://webhook.MATRIX-DOMAIN/webhook/
  userIdPrefix: _webhook_
  allowJsTransformationFunctions: false
  waitForComplete: false
logging:
  colorize: true
  json: false
  level: info
  timestampFormat: HH:mm:ss:SSS
passFile: /data/passkey.pem
cache:
  redisUri: "redis://default:REDIS-PASS@hookshot-redis-master:6379/0"
encryption:
  storagePath: /persistent
bot:
  displayname: Hookshot
```
</details>

Hookshot works perfectly in E2EE rooms as long as all devices/users are in the room first. However, once I add a new device/user, that one can never decrypt the messages. I always get `Unable to decrypt message`

```json
{
  "type": "m.room.message",
  "content": {
    "msgtype": "m.bad.encrypted",
    "body": "** Unable to decrypt: DecryptionError: The sender's device has not sent us the keys for this message. **"
  }
}
```

I'm sorry if I misunderstood what your issue was though. I thought it was the same.

**Comment by PseudoResonance** [comment]
> My issue involved users “losing” their sessions, which led to an “unable to decrypt” error. The only solution was to create a new webhook.

Sorry, I worded it poorly, but I meant that existing devices/users could always decrypt the messages, however any new ones could not. A device that could decrypt it previously never loses the ability to, unless it's logged out and back in, creating a new session.

Personally, I'm just using the Helm chart included in this repo to deploy.

As I said in the reply to jdaviescoates earlier, from my poor understanding, Matrix clients are supposed to listen to new session events (which are the same for a new user, or an existing user logging in on a new device?), and share the current Megolm keys with it. I don't know a lot about the inner workings of the encryption, so I probably misunderstood, but I wasn't able to find anywhere in the Hookshot code that does this and thus figured it's probably not the fault of my config? Although I'd love to be able to just change something and it works haha.

**Comment by CybotTM** [comment]
The problem is back for our users.

It seems that just some update/restart seems partially solved the problem.
Anyway, the problem is still there.
New sessions cannot decrypt messages the user can decrypt on "old" sessions.

**Comment by lephisto** [comment]
Same Issue. Any news?

**Comment by Half-Shot** [comment]
It's a known issue that the hookshot E2EE system is a little shaky due to the upstream library being also quite unstable. The work to fix this will ultimately be part of https://github.com/matrix-org/matrix-hookshot/issues/1001
